### PR TITLE
Minor Changes

### PR DESCRIPTION
**Pull request by 0ddlyoko**
# Description
Some minor changes

# Checklist
The following items should be checked before the pull request can be merged.
- [ ] Version number updated in `settings.gradle`.
- [x] Unit tests added.
- [x] Code follows existing code format.

### CODE REVIEW

File: src/main/java/be/seeseemelk/mockbukkit/attribute/AttributeInstanceMock.java
Comment (TheBusyBiscuit) [inline src/main/java/be/seeseemelk/mockbukkit/attribute/AttributeInstanceMock.java:6]: unused imports

File: src/main/java/be/seeseemelk/mockbukkit/block/data/BlockDataMock.java
Comment (TheBusyBiscuit) [inline src/main/java/be/seeseemelk/mockbukkit/block/data/BlockDataMock.java:3]: Unused import

File: src/main/java/be/seeseemelk/mockbukkit/entity/ArmorStandMock.java
Comment (TheBusyBiscuit) [inline src/main/java/be/seeseemelk/mockbukkit/entity/ArmorStandMock.java:49]: Inconsistently placed annotations, on all methods you added it inline before the return type, here you added it to the method body.

File: src/main/java/be/seeseemelk/mockbukkit/entity/EntityEquipmentMock.java
Comment (TheBusyBiscuit) [inline src/main/java/be/seeseemelk/mockbukkit/entity/EntityEquipmentMock.java:102]: This implementation is redundant.
There is no point in storing the variable when we can just return it immediately.

File: src/main/java/be/seeseemelk/mockbukkit/entity/EntityEquipmentMock.java
Comment (TheBusyBiscuit) [inline src/main/java/be/seeseemelk/mockbukkit/entity/EntityEquipmentMock.java:241]: No annotation here?

File: src/main/java/be/seeseemelk/mockbukkit/entity/EntityEquipmentMock.java
Comment (TheBusyBiscuit) [inline src/main/java/be/seeseemelk/mockbukkit/entity/EntityEquipmentMock.java:248]: else?

File: src/main/java/be/seeseemelk/mockbukkit/entity/EntityMock.java
Comment (TheBusyBiscuit) [inline src/main/java/be/seeseemelk/mockbukkit/entity/EntityMock.java:386]: You should add a message for context to the notNull() call

File: src/main/java/be/seeseemelk/mockbukkit/entity/LivingEntityMock.java
Comment (TheBusyBiscuit) [inline src/main/java/be/seeseemelk/mockbukkit/entity/LivingEntityMock.java:14]: Never use wildcard imports please.

File: src/main/java/be/seeseemelk/mockbukkit/inventory/InventoryMock.java
Comment (TheBusyBiscuit) [inline src/main/java/be/seeseemelk/mockbukkit/inventory/InventoryMock.java:276]: Please add a message for context.

File: src/main/java/be/seeseemelk/mockbukkit/inventory/InventoryMock.java
Comment (TheBusyBiscuit) [inline src/main/java/be/seeseemelk/mockbukkit/inventory/InventoryMock.java:296]: Please add a message for context.

File: src/main/java/be/seeseemelk/mockbukkit/inventory/meta/BookMetaMock.java
Comment (TheBusyBiscuit) [inline src/main/java/be/seeseemelk/mockbukkit/inventory/meta/BookMetaMock.java:134]: Please add a message for context.

File: src/main/java/be/seeseemelk/mockbukkit/inventory/meta/EnchantedBookMetaMock.java
Comment (TheBusyBiscuit) [inline src/main/java/be/seeseemelk/mockbukkit/inventory/meta/EnchantedBookMetaMock.java:76]: Please add a message for context.

File: src/main/java/be/seeseemelk/mockbukkit/inventory/meta/EnchantedBookMetaMock.java
Comment (TheBusyBiscuit) [inline src/main/java/be/seeseemelk/mockbukkit/inventory/meta/EnchantedBookMetaMock.java:93]: Please add a message for context.

File: src/main/java/be/seeseemelk/mockbukkit/inventory/meta/EnchantedBookMetaMock.java
Comment (TheBusyBiscuit) [inline src/main/java/be/seeseemelk/mockbukkit/inventory/meta/EnchantedBookMetaMock.java:88]: Revert this please.
```suggestion
		return prev == null || prev.intValue() != level;
```

File: src/main/java/be/seeseemelk/mockbukkit/inventory/meta/KnowledgeBookMetaMock.java
Comment (TheBusyBiscuit) [inline src/main/java/be/seeseemelk/mockbukkit/inventory/meta/KnowledgeBookMetaMock.java:104]: Please add a message for context.

File: src/main/java/be/seeseemelk/mockbukkit/WorldMock.java
Comment (TheBusyBiscuit) [inline src/main/java/be/seeseemelk/mockbukkit/WorldMock.java:448]: Please add a message for context.

File: src/main/java/be/seeseemelk/mockbukkit/WorldMock.java
Comment (TheBusyBiscuit) [inline src/main/java/be/seeseemelk/mockbukkit/WorldMock.java:449]: Please add a message for context.

File: src/main/java/be/seeseemelk/mockbukkit/WorldMock.java
Comment (TheBusyBiscuit) [inline src/main/java/be/seeseemelk/mockbukkit/WorldMock.java:1086]: Please add a message for context.

File: src/main/java/be/seeseemelk/mockbukkit/WorldMock.java
Comment (TheBusyBiscuit) [inline src/main/java/be/seeseemelk/mockbukkit/WorldMock.java:1087]: Please add a message for context.

File: src/main/java/be/seeseemelk/mockbukkit/WorldMock.java
Comment (TheBusyBiscuit) [inline src/main/java/be/seeseemelk/mockbukkit/WorldMock.java:1113]: Please add a message for context.

File: src/main/java/be/seeseemelk/mockbukkit/entity/ArmorStandMock.java
Comment (TheBusyBiscuit) [inline src/main/java/be/seeseemelk/mockbukkit/entity/ArmorStandMock.java:59]: Why?

File: src/main/java/be/seeseemelk/mockbukkit/entity/ArmorStandMock.java
Comment (TheBusyBiscuit) [inline src/main/java/be/seeseemelk/mockbukkit/entity/ArmorStandMock.java:73]: Why?

File: src/main/java/be/seeseemelk/mockbukkit/entity/ArmorStandMock.java
Comment (TheBusyBiscuit) [inline src/main/java/be/seeseemelk/mockbukkit/entity/ArmorStandMock.java:87]: Why?

File: src/main/java/be/seeseemelk/mockbukkit/entity/ArmorStandMock.java
Comment (TheBusyBiscuit) [inline src/main/java/be/seeseemelk/mockbukkit/entity/ArmorStandMock.java:101]: Why?

**Review by TheBusyBiscuit — CHANGES_REQUESTED** [review]
I didn't review every single file yet.
But I already found of things which will need to be fixed before I'd feel comfortable merging this.
Also I didn't check but I'll trust you that the annotations are all inline with the method contracts of the Bukkit API.

Another very important thing to note is that Jetbrains annotations do not validate parameters, so you may need to make sure that the appropriate exception `IllegalArgumentException` or `NullPointerException` is thrown for null values.

**Comment by 0ddlyoko** [comment]
> Another very important thing to note is that Jetbrains annotations do not validate parameters, so you may need to make sure that the appropriate exception `IllegalArgumentException` or `NullPointerException` is thrown for null values.

Yep, I know for that.
This is why we should use Validate.notNull(...) everywhere

File: src/main/java/be/seeseemelk/mockbukkit/entity/ArmorStandMock.java
Comment (0ddlyoko) [reply src/main/java/be/seeseemelk/mockbukkit/entity/ArmorStandMock.java:59]: getBoots() could return null, getItem(...) not. Another way is to check if getBoots() returns null. If true, return Air

File: src/main/java/be/seeseemelk/mockbukkit/entity/EntityEquipmentMock.java
Comment (0ddlyoko) [reply src/main/java/be/seeseemelk/mockbukkit/entity/EntityEquipmentMock.java:241]: We don't know what to do if we give a different array size. Should we add the first 4 items ? Should we do nothing ? Should we throw an Exception ?

File: src/main/java/be/seeseemelk/mockbukkit/entity/EntityEquipmentMock.java
Comment (0ddlyoko) [reply src/main/java/be/seeseemelk/mockbukkit/entity/EntityEquipmentMock.java:248]: We don't know what to do if we give a different array size. Should we add the first 4 items ? Should we do nothing ? Should we throw an Exception ?

File: src/main/java/be/seeseemelk/mockbukkit/entity/LivingEntityMock.java
Comment (0ddlyoko) [reply src/main/java/be/seeseemelk/mockbukkit/entity/LivingEntityMock.java:14]: Yes sorry, I'll disable this feature in Intellij

File: src/main/java/be/seeseemelk/mockbukkit/inventory/meta/EnchantedBookMetaMock.java
Comment (0ddlyoko) [reply src/main/java/be/seeseemelk/mockbukkit/inventory/meta/EnchantedBookMetaMock.java:88]: Why ?
https://stackoverflow.com/a/11490442

File: src/main/java/be/seeseemelk/mockbukkit/entity/EntityEquipmentMock.java
Comment (TheBusyBiscuit) [reply src/main/java/be/seeseemelk/mockbukkit/entity/EntityEquipmentMock.java:248]: We certainly know what to do.
MockBukkit is a **mock** of CraftBukkit after all, so all methods should always reflect the implementations found in CraftBukkit, for this method the standard implementation is as follows:

```java
    @Override
    public void setArmorContents(final ItemStack[] items) {
        this.setEquipment(EnumItemSlot.FEET, (items.length >= 1) ? items[0] : null);
        this.setEquipment(EnumItemSlot.LEGS, (items.length >= 2) ? items[1] : null);
        this.setEquipment(EnumItemSlot.CHEST, (items.length >= 3) ? items[2] : null);
        this.setEquipment(EnumItemSlot.HEAD, (items.length >= 4) ? items[3] : null);
    }
```

Not only did you get the order completely reversed, but smaller arrays are completely ignored, whereas they should be allowed.

**Comment by TheBusyBiscuit** [comment]
> Yep, I know for that.
> This is why we should use Validate.notNull(...) everywhere

We shouldn't.
I would agree if this was standalone software, but this is a **mock**, so we should only throw an `IllegalArgumentException` where CraftBukkit does.
There are a lot of methods in CB that will also just result in a `NullPointerException`.

**Comment by 0ddlyoko** [comment]
Rip

File: src/main/java/be/seeseemelk/mockbukkit/inventory/meta/EnchantedBookMetaMock.java
Comment (TheBusyBiscuit) [reply src/main/java/be/seeseemelk/mockbukkit/inventory/meta/EnchantedBookMetaMock.java:88]: You should perhaps read the thread you posted.
Because it precisely tells you why this is wrong...

The `==` operator is **not** an equality check, it is a memory address comparison.
Only for primitives can it be used to check for equality but not for Objects. The `Integer` class is what we call a "boxed primitive", it's a wrapper for a primitive but itself is still an `Object` and will follow the object's rules.
Two objects will only ever be equal if they are stored under the same memory location.
Now there are some exceptions due to JVM optimizations which will for example put literals and constants under the same memory location sometimes. But for almost all production code the usage of `==` for comparing Objects should be avoided **at all cost**. It is a really bad practice. And that also applies for boxed integers.
The JVM may still do an implict type transformation to make this work sometimes but it is inherently flawed and should not be relied upon.
That's why this change needs to reverted immediately, we need to make sure that this is correctly converted into a primitive.

**Comment by seeseemelk** [comment]
I'm afraid I cannot merge this pull request for a couple of reasons. The PR description simply says "Some minor changes", which is not very descriptive.

If it was a small PR that only changed a couple of files, this would perhaps be okay. However, this PR changes 975 lines across 71 files. This is in no way 'minor', and it is not really reviewable at all.

The commit history does not make it any easier to try and understand it, and it also does not allow me to easily split it up into smaller (manageable) PRs. 

If the description was clearer about why it was so large, this *might* have been passable, but in it's current state I don't even want to begin reviewing this.

I'm closing this PR. If you still want to see this getting merged, you can split this PR up into several smaller pull requests which can be reviewed and merged individually. This also means that if a small change blocks a certain PR, other unrelated changes are not blocked (as they should be different PRs).

**Pull request closed without merging**
